ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/FollowerPromptness.hs

```diff
@@ -0,0 +1,214 @@
+              case ev of
+                ChainDBImpl.PipeliningEvent pev -> case pev of
+                  ChainDBImpl.SetTentativeHeader hdr FallingEdge -> do
+                    addTiming varTentativeSetTimings (headerHash hdr)
+                    -- Wait 1s to delay the further chain selection logic (i.e.
+                    -- simulating expensive block body validation).
+                    threadDelay artificialDelay
+                  ChainDBImpl.OutdatedTentativeHeader hdr ->
+                    addTiming varTentativeUnsetTimings (headerHash hdr)
+                  ChainDBImpl.TrapTentativeHeader hdr ->
+                    addTiming varTentativeUnsetTimings (headerHash hdr)
+                  _ -> pure ()
+                _ -> pure ()
+            _ -> pure ()
+    chainDB <- openChainDB registry chainDBTracer
+
+    -- Continually fetch instructions from a tentative follower.
+    follower <-
+      ChainDB.newFollower chainDB registry ChainDB.TentativeChain ChainDB.GetHash
+    _ <- forkLinkedThread registry "Follower listener" $ forever $
+      ChainDB.followerInstructionBlocking follower >>= \case
+       Chain.AddBlock hdrHash -> addTiming varFollowerInstrTimings hdrHash
+       Chain.RollBack _       -> pure ()
+
+    -- Add all blocks to the ChainDB.
+    let addBlock = ChainDB.addBlock_ chainDB Punishment.noPunishment
+    for_ chainUpdates \case
+      AddBlock blk      -> addBlock blk
+      SwitchFork _ blks -> for_ blks addBlock
+
+    tentativeHeaderSetTimings   <- readTVarIO varTentativeSetTimings
+    tentativeHeaderUnsetTimings <- readTVarIO varTentativeUnsetTimings
+    followerInstrTimings        <- readTVarIO varFollowerInstrTimings
+    traceByTime                 <- getTrace
+    pure FollowerPromptnessOutcome {..}
+  where
+    openChainDB ::
+         ResourceRegistry m
+      -> Tracer m (ChainDBImpl.TraceEvent TestBlock)
+      -> m (ChainDB m TestBlock)
+    openChainDB registry cdbTracer = do
+        chainDbArgs <- do
+          let mcdbTopLevelConfig = singleNodeTestConfigWithK securityParam
+              mcdbChunkInfo      = mkTestChunkInfo mcdbTopLevelConfig
+              mcdbInitLedger     = testInitExtLedger
+              mcdbRegistry       = registry
+          mcdbNodeDBs <- emptyNodeDBs
+          let cdbArgs = fromMinimalChainDbArgs MinimalChainDbArgs{..}
+          pure $ cdbArgs { cdbTracer = cdbTracer }
+        (_, (chainDB, ChainDBImpl.Internal{intAddBlockRunner})) <-
+          allocate
+            registry
+            (\_ -> ChainDBImpl.openDBInternal chainDbArgs False)
+            (ChainDB.closeDB . fst)
+        _ <- forkLinkedThread registry "AddBlockRunner" intAddBlockRunner
+        pure chainDB
+
+    withTime = contramapM \ev -> (,ev) <$> getMonotonicTime
+
+    addTiming varTiming hash = do
+        now <- getMonotonicTime
+        atomically $ modifyTVar varTiming $
+          Map.alter (Just . maybe (Set.singleton hash) (Set.insert hash)) now
```

Comment:
Hmm. Would `Just . Set.insert hash . maybe Set.empty Set.singleton` be easier to read?

Or maybe even `Map.unionWith Set.union (Map.singleton now hash)`?

Reply:
I like `Map.unionWith Set.union (Map.singleton now (Set.singleton hash))`, as it does not require looking up what exactly `alter` was doing again.